test/hotspot/jtreg/compiler/c2/TestAddXorIdeal.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright (c) 2021, Alibaba Group Holding Limited. All Rights Reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+/*
+ * @test
+ * @bug 8273021
+ * @summary C2: Improve Add and Xor ideal optimizations
+ * @library /test/lib
+ * @run main/othervm -XX:-TieredCompilation -XX:TieredStopAtLevel=4
+ *                   -XX:CompileCommand=dontinline,compiler.c2.TestAddXorIdeal::test*
+ *                   compiler.c2.TestAddXorIdeal
+ */
+package compiler.c2;
+
+import java.util.Random;
+
+import jdk.test.lib.Asserts;
+
+public class TestAddXorIdeal {
+
+    public static int test1(int x) {
+        return ~x + 1;
+    }
+
+    public static int test2(int x) {
+        return ~(x - 1);
+    }
+
+    public static long test3(long x) {
+        return ~x + 1L;
+    }
+
+    public static long test4(long x) {
+        return ~(x - 1L);
+    }
+
+    public static int test5(int x) {
+        return 1 + ~x;
+    }
+
+    public static int test6(int x) {
+        return ~(-1 + x);
+    }
+
+    public static long test7(long x) {
+        return 1L + ~x;
+    }
+
+    public static long test8(long x) {
+        return ~(-1L + x);
+    }
+
+    public static void main(String... args) {
+        Random random = new Random();
+        int n = 0;
+        long n1 = 0;
```

Comment:
Should be declared in the loop.

Reply:
Do you mean declared within loop body? I've changed but it looks like a perference problem.

Reply:
Yes, in Java, local variables should be declared as close as possible to the point they are first used at (see, for example, [Google's Java Style Guide](https://google.github.io/styleguide/javaguide.html#s4.8.2-variable-declarations)). The declaration does not affect performance.

Here's how I would write the loop to improve readability:
```
for (int j = 0; j < 50_000; j++) {
  int i = random.nextInt();
  long l = random.nextLong();
  Asserts.assertTrue(test1(i) == -i);
  Asserts.assertTrue(test2(i) == -i);
  Asserts.assertTrue(test3(l) == -l);
  ...
```
Summary: No need to use negative initial value for loop induction variable (as it is not even used), increase number of iterations to ensure C2 compilation (you are running without `-Xbatch`), use same random numbers per loop iteration, use more intuitive variable names.

But these are just code style details, feel free to ignore.

Reply:
Thanks for patient. I agree with your comment that using 0 as initial value and increasing iterations. :)

I try to use `a` and `b` as variable names since `long l` looks like `long 1`